startup: go to buffer 2 if stdin is empty

justinmk opened the pull request:
If stdin is not a TTY we read it into buffer 1, as text. But if the
stdin pipe is empty, Nvim was most likely invoked for some other reason.
DWIM: select buffer 2 (if it exists). Example:

    echo file1 | xargs nvim

closes #8560
closes #8561
ref https://github.com/equalsraf/neovim-qt/issues/417

justinmk committed:
startup: go to buffer 2 if stdin is empty

If stdin is not a TTY we read it into buffer 1, as text. But if the
stdin pipe is empty, Nvim was most likely invoked for some other reason.
DWIM: select buffer 2 (if it exists). Example:

    echo file1 | xargs nvim

closes #8560
closes #8561
ref https://github.com/equalsraf/neovim-qt/issues/417

alok commented:
Would this approach work if, say, we used `nvr` to open a file from inside the embedded terminal? Wouldn't the empty buffer not have buffer ID 1?

gibfahn commented:
This will still leave the empty buffer open right? Would it be possible to close it (or not open it at all)?

justinmk commented:
> Would this approach work if, say, we used nvr to open a file from inside the embedded terminal?

is that not already working? This isn't going to affect that anyway.

jamessan commented:
> This will still leave the empty buffer open right?

Yes

> Would it be possible to close it (or not open it at all)?

It has to be opened since that's how we're able to tell the buffer is empty.

This change is simply a heuristic to make `nvim foo` work as expected when `stdin` isn't a tty.

From the perspective of nvim, `echo | nvim foo` is the same as `echo foo | xargs nvim`.  The user can easily tell the difference between the two, but in the former case deleting the initial buffer is hiding useful information (that the command didn't produce any output) from the user.

justinmk committed:
startup: delete empty stdin buffer if other files were opened

DWIM: avoid empty buffer 1 when stdin was empty. If other files were
specified at startup, we assume that stdin is only accidentally
not-a-TTY: user did not intend to send text from it.

ref #8560
ref #8561

gibfahn commented:
>in the former case deleting the initial buffer is hiding useful information (that the command didn't produce any output) from the user.

If the buffer is only deleted if it's empty, then isn't that information inferable from whether the buffer exists?

In any case that seems like useful debugging information, but having an empty open buffer every time neovim opens is a bit of a pain.

A lot of scripts do `<command> | xargs $VISUAL`, which means that they will now have the open buffer. For my own shortcuts I could do something like `$VISUAL $(<command>)`, but then you run into the max number of positional arguments limit.

jamessan commented:
> A lot of scripts do `<command> | xargs $VISUAL`, which means that they will now have the open buffer.

Sure, but what harm does the open buffer do?

gibfahn commented:
>Sure, but what harm does the open buffer do?

Fair question, I know a lot of people don't keep track of buffers, but if you use a buffer tabline plugin like [buftabline](https://github.com/ap/vim-buftabline), [vim-bufferline](https://github.com/bling/vim-bufferline) or [wintabs](https://github.com/zefei/vim-wintabs), then it means you always have this annoying extra tab.

This may not seem like a big deal, but it's a big enough annoyance to make me switch back to 0.2.1.

![image](https://user-images.githubusercontent.com/15943089/41475745-7b321eb8-70b7-11e8-8395-ada93d0de68c.png)

justinmk commented:
I agree with @jamessan , but my sense is that the empty buffer might cause more confusion that it solves. So latest commit deletes it. We can revert if it causes problems.

> it's a big enough annoyance to make me switch back to 0.2.1.

That throws out quite a lot of improvements in 0.3.0, particularly some major job-control bugs were fixed on 0.3.0, as well as performance improvements and many other details.

gibfahn commented:
>That throws out quite a lot of improvements in 0.3.0, particularly some major job-control bugs were fixed on 0.3.0, as well as performance improvements and many other details.

Yeah, that's exactly my point. I know 0.3.0 has lots of great improvements, which is why I was keen to update. But this change causes enough mental distraction for me to make me prefer 0.2.1.

gibfahn commented:
@justinmk just wanted to say thanks for fixing this, works like a charm in v0.3.1!
